### PR TITLE
feat: add global context documentation

### PR DESCRIPTION
**Pull request by KhaledEmaraDev**
## Related issue #

Add Global Context Documentation

## Related PRs

* https://github.com/kyverno/kyverno/pull/9601
* https://github.com/kyverno/kyverno/pull/9602
* https://github.com/kyverno/kyverno/pull/9615
* https://github.com/kyverno/kyverno/pull/9619
* https://github.com/kyverno/kyverno/pull/9640
* https://github.com/kyverno/kyverno/pull/9652
* https://github.com/kyverno/kyverno/pull/9661
* https://github.com/kyverno/kyverno/pull/9813

## Checklist

- [x] I have read the [contributing guidelines](https://github.com/kyverno/website/blob/main/CONTRIBUTING.md).
- [x] I have inspected the website preview for accuracy.
- [x] I have signed off my issue.

### CODE REVIEW

**Comment by chipzoller** [comment]
Please link to the PR(s) in the code repo responsible for this. And based on yesterday's discussion, it seems like this PR isn't yet ready so converting to draft.

File: content/en/docs/Writing policies/external-data-sources.md [hunk 1]
```diff
@@ -589,6 +589,74 @@ spec:
...
+**Note:** The data returned by Global Context Entries may vary depending on whether it's a Kubernetes resource or an API call. Consequently, the JMESPath expression used to manipulate the data may differ as well.
+
+* For Kubernetes Resources, to obtain the length of the array, use length(@).
+* For API Calls returning a list, you can use items | length(@) to get the length.
+
+Ensure to use the appropriate JMESPath expression based on the type of data being accessed to ensure accurate processing within policies.
+
+### Notes:
+
+If you want to cache resources on all namespaces, leave the namespace  empty.
```
Comment (realshuting) [inline content/en/docs/Writing policies/external-data-sources.md:658]:
```suggestion
If you want to cache resources on all namespaces, leave the namespace empty.
```

File: content/en/docs/Writing policies/external-data-sources.md [hunk 2]
```diff
@@ -589,6 +589,74 @@ spec:
             value: false
 ```
 
+## Global Context
+
+Global Context is a feature introduced in Kyverno that allows users to cache Kubernetes resources or external API calls for later reference within policies. It provides a mechanism to efficiently retrieve and utilize data across policies, enhancing flexibility and performance in policy enforcement.
+
+### Global Context Entry:
+
+Imagine declaring resource or API call data needs once, then easily referencing it throughout your policies. That's what Kyverno's new GlobalContextEntry custom resource type offers. You can create dedicated entries specifying:
```
Comment (realshuting) [inline content/en/docs/Writing policies/external-data-sources.md:598]: Can we clarify the scope of a GlobalContextEntry?

File: content/en/docs/Writing policies/external-data-sources.md [hunk 3]
```diff
@@ -589,6 +589,74 @@ spec:
...
+
+```yaml
+apiVersion: kyverno.io/v2alpha1
+kind: GlobalContextEntry
+metadata:
+  name: deployments
+spec:
+  apiCall:
+    urlPath: "/apis/apps/v1/namespaces/test-globalcontext/deployments"
+    refreshInterval: 10s
```
Comment (realshuting) [inline content/en/docs/Writing policies/external-data-sources.md:635]: Can we explain `refreshInterval` in details?

File: content/en/docs/Writing policies/external-data-sources.md [hunk 4]
```diff
@@ -589,6 +589,74 @@ spec:
...
+  kubernetesResource:
+    group: apps
+    version: v1
+    resource: deployments
+    namespace: test-globalcontext
+```
+
+##### API Call
+
+Same but with apiCall:
```
Comment (realshuting) [inline content/en/docs/Writing policies/external-data-sources.md:625]:
```suggestion
Same but with `apiCall`:
```

**Review by realshuting — CHANGES_REQUESTED** [review]

**Review by chipzoller — CHANGES_REQUESTED** [review]
Couple things.

1. Don't use punctuation in Markdown headings. This causes problems with anchors for links.
2. You don't need all these various headings within the parent "Global Context". I would just go with the parent.

**Comment by KhaledEmaraDev** [comment]
Thanks @chipzoller, removed headers. Also added PR links.

File: content/en/docs/Writing policies/external-data-sources.md [hunk 5]
```diff
@@ -589,6 +589,68 @@ spec:
...
+
+Global Context is a feature introduced in Kyverno that allows users to cache Kubernetes resources or external API calls for later reference within policies. It provides a mechanism to efficiently retrieve and utilize data across policies, enhancing flexibility and performance in policy enforcement.
+
+Imagine declaring resource or API call data needs once, then easily referencing it throughout your policies. That's what Kyverno's new GlobalContextEntry custom resource type offers. Global Context Entries are declared globally using a new CRD irrespective of any Policies you might have and then later referenced as part of a policies's context. There are two types of Global Context Entries:
+
+1. **Kubernetes Resource:** Reference any Kubernetes resource (optionally namespace) to pre-populate the cache with its latest state.
+
+**Example:**
+
+This `CachedContextEntry` caches all `Deployment` resources in the `test-globalcontext` namespace:
```
Comment (realshuting) [inline content/en/docs/Writing policies/external-data-sources.md:602]: What is this `CachedContextEntry`?

File: content/en/docs/Writing policies/external-data-sources.md [hunk 6]
```diff
@@ -589,6 +589,68 @@ spec:
...
+  name: deployments
+spec:
+  kubernetesResource:
+    group: apps
+    version: v1
+    resource: deployments
+    namespace: test-globalcontext
+```
+
+*Notes:* Leave namespacce empty for cluster scoped resources.
```
Comment (vishal-chdhry) [inline content/en/docs/Writing policies/external-data-sources.md:617]:
```suggestion
*Note:* Leave namespacce empty for cluster scoped resources.
```

File: content/en/docs/Writing policies/external-data-sources.md [hunk 7]
```diff
@@ -589,6 +589,68 @@ spec:
...
+kind: GlobalContextEntry
+metadata:
+  name: deployments
+spec:
+  apiCall:
+    urlPath: "/apis/apps/v1/namespaces/test-globalcontext/deployments"
+    refreshInterval: 10s
+```
+
+*Notes*: `APICall` Global Context Entries are implemented by periodically making a call to the `urlPath` every `refreshInterval`, thus beware of stale data.
```
Comment (vishal-chdhry) [inline content/en/docs/Writing policies/external-data-sources.md:636]:
```suggestion
*Note:* `APICall` Global Context Entries are implemented by periodically making a call to the `urlPath` every `refreshInterval`, thus beware of stale data.
```

File: content/en/docs/Writing policies/external-data-sources.md [hunk 8]
```diff
@@ -589,6 +589,68 @@ spec:
...
+    globalReference:
+      name: deployments
+```
+
+*Note:* The data returned by Global Context Entries may vary depending on whether it's a Kubernetes resource or an API call. Consequently, the JMESPath expression used to manipulate the data may differ as well.
+
+* For Kubernetes Resources, to obtain the length of the array, use length(@).
+* For API Calls returning a list, you can use items | length(@) to get the length.
+
+Ensure to use the appropriate JMESPath expression based on the type of data being accessed to ensure accurate processing within policies.
```
Comment (vishal-chdhry) [inline content/en/docs/Writing policies/external-data-sources.md:652]: I dont think `Ensure to use` is grammatically correct, it just sounds off to me 🤔, cc @chipzoller

Comment (chipzoller) [reply content/en/docs/Writing policies/external-data-sources.md:652]: Technically correct, but "ensure you use" is a little more common.

Avoid using title case for nouns in English which aren't proper. "Resources" and "Calls" are not proper nouns.

**Review by chipzoller — APPROVED** [review]
Thank you for this PR. Going to merge and work on in `main`.

**Review by realshuting — APPROVED** [review]

**Pull request merged**